x/rollapp/keeper/liveness.go

```diff
@@ -0,0 +1,153 @@
+It will trigger slash/jail operations through the x/sequencers module, at intervals decided by parameters.
+See ADR for more info https://www.notion.so/dymension/ADR-x-Sequencer-Liveness-Slash-Phase-1-5131b4d557e34f4498855831f439d218
+*/
+
+// NextSlashOrJailHeight returns the next height on the HUB to slash or jail the rollapp
+// It will respect all parameters passed in.
+func NextSlashOrJailHeight(
+	hubBlockInterval time.Duration, // average time between hub blocks
+	slashTimeNoUpdate time.Duration, // time until first slash if not updating
+	slashInterval time.Duration, // gap between slash if still not updating
+	jailTime time.Duration, // time until jail if not updating
+	heightHub int64, // current hub height
+	heightLastRollappUpdate int64, // when was the rollapp last updated
+) (
+	heightEvent int64, // hub height to schedule event
+	isJail bool, // is it a jail event? (false -> slash)
+) {
+	// how long has the rollapp been down already?
+	downTime := time.Duration(heightHub-heightLastRollappUpdate) * hubBlockInterval
+	// when should we schedule the next slash/jail, in terms of down time duration?
+	targetDuration := slashTimeNoUpdate
+	if slashTimeNoUpdate < downTime {
+		targetDuration += ((downTime-slashTimeNoUpdate)/slashInterval + 1) * slashInterval
+	}
+	heightEvent = heightLastRollappUpdate + int64((targetDuration+hubBlockInterval-1)/hubBlockInterval)
+	isJail = jailTime <= targetDuration
+	return
+}
+
+// CheckLiveness will slash or jail any sequencers for whom their rollapp has been down
+// and a slash or jail event is due
+func (k Keeper) CheckLiveness(ctx sdk.Context) {
+	h := ctx.BlockHeight()
+	events := k.GetLivenessEvents(ctx, &h)
+	for _, e := range events {
+		err := osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+			return k.HandleLivenessEvent(ctx, e)
+		})
+		if err != nil {
+			k.Logger(ctx).Error(
+				"Check liveness event",
+				"event", e,
+				"err", err,
+			)
+		}
+	}
+}
+
+// HandleLivenessEvent will slash or jail and then schedule a new event in the future.
+func (k Keeper) HandleLivenessEvent(ctx sdk.Context, e types.LivenessEvent) error {
+	if e.IsJail {
+		err := k.sequencerKeeper.JailLiveness(ctx, e.RollappId)
+		if err != nil {
+			return errorsmod.Wrap(err, "jail liveness")
+		}
+	} else {
+		err := k.sequencerKeeper.SlashLiveness(ctx, e.RollappId)
+		if err != nil {
+			return errorsmod.Wrap(err, "slash liveness")
+		}
+	}
+
+	ra := k.MustGetRollapp(ctx, e.RollappId)
+	k.DelLivenessEvents(ctx, ra.LivenessEventHeight, ra.RollappId)
+	k.ScheduleLivenessEvent(ctx, &ra)
+	k.SetRollapp(ctx, ra)
+	return nil
+}
+
+// IndicateLiveness will reschedule pending liveness events to a later date.
+// Modifies the passed in rollapp object.
+func (k Keeper) IndicateLiveness(ctx sdk.Context, ra *types.Rollapp) {
+	ra.LastStateUpdateHeight = ctx.BlockHeight()
+	k.DelLivenessEvents(ctx, ra.LivenessEventHeight, ra.RollappId)
+	k.ScheduleLivenessEvent(ctx, ra)
```

Comment:
so this "reschedule" of the liveness event happens every `UpdateState` and every `EndBlock` for each rollapp. I commented on the ADR how I don't see how this is scalable.

Instead, you could simply compare the last state CreationHeight with the current height, every `UpdateState` and maybe on every epoch, for good measure, and if it's AWOL, start slashing the sequencer

Reply:
resolved offline